shell/platform/android/io/flutter/util/PathUtils.java

```diff
@@ -17,10 +21,19 @@ public static String getDataDirectory(Context applicationContext) {
   }
 
   public static String getCacheDirectory(Context applicationContext) {
+    File cacheDir;
     if (Build.VERSION.SDK_INT >= 21) {
-      return applicationContext.getCodeCacheDir().getPath();
+      cacheDir = applicationContext.getCodeCacheDir();
     } else {
-      return applicationContext.getCacheDir().getPath();
+      cacheDir = applicationContext.getCacheDir();
+    }
+    if (cacheDir == null) {
+      Log.w(TAG, "Failed to get internal cache directory, fallback to get external cache dir.");
+      cacheDir = applicationContext.getExternalCacheDir();
+    }
+    if (cacheDir == null) {
+      throw new RuntimeException("Failed to get cache directory.");
```

Comment:
We should definitely throw something better than an NPE, but I think we should probably throw a FileNotFoundException here rather than the more generic RuntimeException.

Reply:
But FileNotFoundException is IOException and it's a checked exception type, how should I change it? @dnfield

Reply:
I'm not sure if this is a good idea or not, but could we just do something like

```java
// This can happen if the disk is full. This code path is used to set up dart:io's 
// `Directory.systemTemp`. It's unkonwn if the application will ever try to
// use that or not, so do not throw here. In this case, this directory does
// not exist because the disk is full, and the application will later get an
// exception when it tries to actually write.
return applicationContext.getDataDir().getPath() + "/cache";
```